contracts/KyberOasisReserve.sol

```diff
@@ -0,0 +1,273 @@
+
+        tradeToken.approve(otc, 0);
+        _tradeToken.approve(otc, 2**255);
+
+        tradeToken = _tradeToken;
+        TradeTokenSet(tradeToken);
+    }
+
+    event FeeBpsSet(uint feeBps);
+
+    function setFeeBps(uint _feeBps) public onlyAdmin {
+        require(_feeBps < 10000);
+
+        feeBps = _feeBps;
+        FeeBpsSet(feeBps);
+    }
+
+    function getDestQty(ERC20 src, ERC20 dest, uint srcQty, uint rate) public view returns(uint) {
+        uint dstDecimals = getDecimals(dest);
+        uint srcDecimals = getDecimals(src);
+
+        return calcDstQty(srcQty, srcDecimals, dstDecimals, rate);
+    }
+
+    function valueAfterReducingFee(uint val) public view returns(uint) {
+        require(val <= MAX_QTY);
+        return ((10000 - feeBps) * val) / 10000;
+    }
+
+    function getConversionRate(ERC20 src, ERC20 dest, uint srcQty, uint blockNumber) public view returns(uint) {
+        uint  rate;
+        uint  destQty;
+        uint  actualSrcQty;
+        ERC20 wrappedSrc;
+        ERC20 wrappedDest;
+
+        blockNumber;
+
+        if (!tradeEnabled) return 0;
+        if ((tradeToken != src) && (tradeToken != dest)) return 0;
+
+        if (src == ETH_TOKEN_ADDRESS) {
+            wrappedSrc = wethToken;
+            wrappedDest = dest;
+            actualSrcQty = srcQty;
+        } else if (dest == ETH_TOKEN_ADDRESS) {
+            wrappedSrc = src;
+            wrappedDest = wethToken;
+
+            if (srcQty < ETH_SRC_AMOUNT) {
+                /* Assuming token is stable, use a minimal amount to get rate also for small token quant. */
+                actualSrcQty = ETH_SRC_AMOUNT;
+            } else {
+                actualSrcQty = srcQty;
+            }
+        } else {
+            return 0;
+        }
+
+        destQty = otc.getBuyAmount(wrappedDest, wrappedSrc, actualSrcQty);
+        rate = calcRateFromQty(actualSrcQty, valueAfterReducingFee(destQty), MAX_DECIMALS, MAX_DECIMALS);
+
+        return rate;
+    }
+
+    function doTrade(
+        ERC20 srcToken,
+        uint srcAmount,
+        ERC20 destToken,
+        address destAddress,
+        uint conversionRate,
+        bool validate
+    )
+        internal
+        returns(bool)
+    {
+        require((ETH_TOKEN_ADDRESS == srcToken) || (ETH_TOKEN_ADDRESS == destToken));
+        require((tradeToken == srcToken) || (tradeToken == destToken));
+
+        uint actualDestAmount;
+
+        // can skip validation if done at kyber network level
+        if (validate) {
+            require(conversionRate > 0);
+            if (srcToken == ETH_TOKEN_ADDRESS)
+                require(msg.value == srcAmount);
+            else
+                require(msg.value == 0);
+        }
+
+        uint destAmount = getDestQty(srcToken, destToken, srcAmount, conversionRate);
+        // sanity check
+        require(destAmount > 0);
+
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            wethToken.deposit.value(msg.value)();
```

Comment:
does it has a return value? if so, need to check it (`require`)

Reply:
does not return a value.